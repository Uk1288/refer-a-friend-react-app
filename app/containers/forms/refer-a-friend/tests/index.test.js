import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { browserHistory } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from '../../../../configureStore';
import ConnectedReferAFriendForm, {
  ReferAFriendForm,
  mapDispatchToProps,
  mapStateToProps,
} from '../index';
import { rootStateMock } from '../../../../mocks/root.mock';
import {
  referralEmailSuccessStateMock,
  referralEmailPendingStateMock,
  referralEmailErrorStateMock,
  uiStateMock,
} from '../../../../mocks/ui.mock';
import {
  sendReferralEmail,
  resetReferralEmail,
} from '../../../../store/ui/ui.action';

const mockEmail = 'abc@abc.abc';

describe('<ReferAFriendForm />', () => {
  describe('<ConnectedReferAFriendForm />', () => {
    let store;

    beforeAll(() => {
      store = configureStore(rootStateMock, browserHistory);
    });

    it('should render and match the snapshot', () => {
      const {
        container: { firstChild },
      } = render(
        <Provider store={store}>
          <IntlProvider locale="en">
            <ConnectedReferAFriendForm />
          </IntlProvider>
        </Provider>,
      );
      expect(firstChild).toMatchSnapshot();
    });
  });

  describe('<ReferAFriendForm />', () => {
    const defaultProps = {
      handleEmailReferral: jest.fn(),
      referralIsPending: false,
      referralSucceeded: false,
      referralFailed: false,
      resetEmailReferralStatus: jest.fn(),
      referralErrorMsg: 'Error occured',
    };

    it('should render api send email errors', () => {
      const { getByText } = render(
        <IntlProvider locale="en">
          <ReferAFriendForm {...{ ...defaultProps, referralFailed: true }} />
        </IntlProvider>,
      );
      expect(getByText(defaultProps.referralErrorMsg)).toBeTruthy();
    });

    it('should render success message', () => {
      const { getByText } = render(
        <IntlProvider locale="en">
          <ReferAFriendForm {...{ ...defaultProps, referralSucceeded: true }} />
        </IntlProvider>,
      );
      expect(getByText('Referral email sent!')).toBeTruthy();
    });
  });

  describe('mapDispatchToProps', () => {
    const dispatch = jest.fn();
    const dispatchToProps = mapDispatchToProps(dispatch);
    const mockPayload = { email: mockEmail };

    it('handleEmailReferral', () => {
      dispatchToProps.handleEmailReferral(mockPayload);
      expect(dispatch).toHaveBeenCalledWith(sendReferralEmail(mockPayload));
    });

    it('resetEmailReferralStatus', () => {
      dispatchToProps.resetEmailReferralStatus();
      expect(dispatch).toHaveBeenCalledWith(resetReferralEmail());
    });
  });

  describe('mapStateToProps', () => {
    it('referral pending', () => {
      const rootState = {
        ...rootStateMock,
        ui: { ...uiStateMock, referralEmail: referralEmailPendingStateMock },
      };
      const wrapperState = mapStateToProps(rootState);
      expect(wrapperState).toMatchSnapshot();
    });

    it('referral error', () => {
      const rootState = {
        ...rootStateMock,
        ui: { ...uiStateMock, referralEmail: referralEmailSuccessStateMock },
      };
      const wrapperState = mapStateToProps(rootState);
      expect(wrapperState).toMatchSnapshot();
    });

    it('referral success', () => {
      const rootState = {
        ...rootStateMock,
        ui: { ...uiStateMock, referralEmail: referralEmailErrorStateMock },
      };
      const wrapperState = mapStateToProps(rootState);
      expect(wrapperState).toMatchSnapshot();
    });
  });
});
