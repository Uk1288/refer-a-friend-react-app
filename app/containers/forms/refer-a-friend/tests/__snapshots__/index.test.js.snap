// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReferAFriendForm /> <ConnectedReferAFriendForm /> should render and match the snapshot 1`] = `
<div>
  <form>
    <div
      class="refer-a-friend__StyledEmailText-sc-16r8705-0 hDQkIh"
    >
      <span>
        Enter email address
      </span>
    </div>
    <div
      class="refer-a-friend__StyledInputBox-sc-16r8705-1 Qzlag"
    >
      <input
        class="input__StyledInput-sc-10zxn30-0 lkoULL"
        id="email"
        name="email"
        value=""
      />
      <button
        class="button__StyledButton-sc-1bg6su4-0 hygGZF"
        type="submit"
      >
        <span>
          Send
        </span>
      </button>
    </div>
  </form>
</div>
`;

exports[`<ReferAFriendForm /> mapStateToProps referral error 1`] = `
Object {
  "referralErrorMsg": undefined,
  "referralFailed": false,
  "referralIsPending": false,
  "referralSucceeded": true,
}
`;

exports[`<ReferAFriendForm /> mapStateToProps referral pending 1`] = `
Object {
  "referralErrorMsg": undefined,
  "referralFailed": false,
  "referralIsPending": true,
  "referralSucceeded": false,
}
`;

exports[`<ReferAFriendForm /> mapStateToProps referral success 1`] = `
Object {
  "referralErrorMsg": "Error occured",
  "referralFailed": true,
  "referralIsPending": false,
  "referralSucceeded": false,
}
`;
